### ISSUE TITLE
Migrate away from Travis CI

### ISSUE THREAD

**Issue by aothms** [IfcOpenShell/IfcOpenShell#1625]
Re https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing we had been been using their service quite generously, but our commit rate would make paying for the product quite expensive. We already have Github actions in place, we just need one for Pull Requests.

**Comment by Krande** [comment]
If you haven't started on this already, I can give this a shot. What functionality specifically do you want for each pull request?

**Comment by aothms** [comment]
That'd be really great! Most important is whether the code compiles. But
Travis did a bit more and ran some checks. Maybe take the Travis yml as a
starting point. Let me know if you have questions

Sent from a mobile device. Excuse my brevity. Kind regards, Thomas

Op ma 9 aug. 2021 09:23 schreef Kristoffer Andersen <
***@***.***>:

**Comment by Krande** [comment]
Sounds good!

I guess we can either use the default ubuntu version of github action and install the prerequisites (which I believe is done for the [release.yml](https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/.github/workflows/release.yml) workflow. 

OR we can create a docker image with all the necessary prerequisites pre-installed and do the compilation and tests inside the container. 

My gut feeling is that the latter option would be faster (depending on what we can do with caching each step). And personally I would love to have docker images with the latest version of ifcopenshell for local testing AND for compilation of daily versions.

Do we have a "latest" docker image for the pre-requisites of ifcopenshell hosted somewhere that I can pull into the github action workflow? If not maybe we should also include a github action workflow to build\update the pre-requisite docker images and maybe host them on github? 

let me know what you think! In the mean time I can start experimenting with building a docker image based on the `travis.yml` file.

**Comment by aothms** [comment]
> use the default ubuntu version of github action and install the prerequisites 

> OR we can create a docker image 

I don't have a strong preference, but thanks to @plaes we have ccache enabled in the ci action which results in quite a massive speedup as most of the commits are python files anyway that don't incur any changes in compilation. https://github.com/IfcOpenShell/IfcOpenShell/commit/184a189f2a8dd4b881876c6fbeb1e5607db99185 That is potentially harder in Docker?

> Do we have a "latest" docker image for the pre-requisites 

Another issue we have is that I think ci action still depends on the old OCE package in Ubuntu. That's OCCT 6.9 or something which means a lot of new additions in OCCT are not available (related to booleans and layersets in particular).

**Comment by Krande** [comment]
Okay, I will see what I come up with :)

I do think docker will be able to achieve speed improvements in similar magnitudes by basically building the image complete with all the pre-requisites and then use caching on the image layers during docker pull. I can test it out on my fork and see if it works as intended and compare speeds . 

Regarding OCC versions. I did a local docker test using `ubuntu:focal` and it seemed to install OCC v7.3. I could also produce an alternative docker image with the latest OCC version as well.

Maybe I could apply a similar approach as was the case in the blenderbim addon makefile (before removal of OCC as a dependency) and just download the OCC versions from conda directly like from this urL: https://anaconda.org/conda-forge/occt/7.5.2/download/linux-64/occt-7.5.2-h7391655_0.tar.bz2?

In the event using github actions directly, I guess this would still work together with a caching mechanism.

I will let you know how it progresses as soon as I have done some tests on my fork.

**Comment by aothms** [comment]
> install OCC v7.3

That's good enough.

> I do think docker will be able to achieve speed improvements in similar magnitudes by basically building the image complete with all the pre-requisites and then use caching on the image layers during docker pull.

It depends. I guess if you somehow manage to do:

```
ADD cpp_sources
RUN make build
ADD python sources
RUN make install
```

then at least the compilation step would be skipped in case of python only changes. But with cccache you can cache more granularly and skip the compilation of cpp source files for which no change was made.

Note that for me runtime of the entire script is not a major concern though.

**Comment by Krande** [comment]
Ah, I see your point. Then I guess a regular github action workflow would be faster. I think I have solved it by manually downloading occt (currently testing with v7.5.2) from conda (and at the same time install tbb using apt). 

Okay, so I have made progress on both docker and using github actions so that they both compile correctly. However I am a bit stuck at how I would install the python package (its currently giving me `no module named ifcopenshell` errors in both my docker build and pure github action build. 

Is there one of the passed keyword arguments that installs the ifcopenshell python package into an existing python package or does it rely on post-copying? Or maybe I have made too many errors when porting over code from the travis.yml file and the existing `ci.yml` file? :)

Excerpt from my github action workflow file:

```
- name: Build ifcopenshell
  run: |
    mkdir build && cd build
    cmake \
     -DCMAKE_C_COMPILER_LAUNCHER=ccache \
     -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
     -DCMAKE_INSTALL_PREFIX=$PWD/install/ \
     -DCMAKE_BUILD_TYPE=Release \
     -DCMAKE_PREFIX_PATH=/usr \
     -DCMAKE_SYSTEM_PREFIX_PATH=/usr \
     -DBUILD_PACKAGE=On \
     -DOCC_INCLUDE_DIR=${{env.tmp_dir}}/include/opencascade \
     -DOCC_LIBRARY_DIR=${{env.tmp_dir}}/lib \
     -DPYTHON_EXECUTABLE:FILEPATH=/usr/bin/python3 \
     -DPYTHON_INCLUDE_DIR:PATH=/usr/include/python3.8 \
     -DPYTHON_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libpython3.8.so \
     -DCOLLADA_SUPPORT=Off \
     -DLIBXML2_INCLUDE_DIR=/usr/include/libxml2 \
     -DLIBXML2_LIBRARIES=/usr/lib/x86_64-linux-gnu/libxml2.so \
     ../cmake
    make -j $(nproc)
    make install
- name: Test
  run: |
    cd test
    /usr/bin/python tests.py
    /usr/bin/python src/ifcopenshell-python/ifcopenshell/test_ids.py
```
and resulting error when testing (taken from this specific workflow https://github.com/Krande/IfcOpenShell/runs/3302074190?check_suite_focus=true ). 
```
Run cd test
Traceback (most recent call last):
  File "tests.py", line 25, in <module>
    import ifcopenshell
ModuleNotFoundError: No module named 'ifcopenshell'
```

let me know if you have any ideas?

**Comment by aothms** [comment]
I only had the tiniest look, will look again tomorrow, but the install path for the ifcopenshell module indeed seems odd to me https://github.com/Krande/IfcOpenShell/runs/3303181567?check_suite_focus=true#step:6:1117

**Comment by aothms** [comment]
Ah now I get it. You have `-DBUILD_PACKAGE=On`. I was a bit slow in figuring it out, this was a byproduct of #1307 which I slightly tweaked in https://github.com/IfcOpenShell/IfcOpenShell/commit/a56ca661e8d2d2052b68eb1af30b3deb3bebd5af If you set it to off the python ifcopenshell lib will be installed into a directory in which the interpreter site-packages dir. If you set it to on, all build artefacts end up in a single folder for packaging.

**Comment by Krande** [comment]
Okay, great! 

Now we are one step closer as `ifcopenshell` is now found :)

However, now I am getting the following errors ([here](https://github.com/Krande/IfcOpenShell/runs/3332262340?check_suite_focus=true#step:8:14))

```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/ifcopenshell/__init__.py", line 43, in <module>
    from . import ifcopenshell_wrapper
  File "/usr/lib/python3/dist-packages/ifcopenshell/ifcopenshell_wrapper.py", line 13, in <module>
    from . import _ifcopenshell_wrapper
ImportError: libTKernel.so.7: cannot open shared object file: No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "tests.py", line 25, in <module>
    import ifcopenshell
  File "/usr/lib/python3/dist-packages/ifcopenshell/__init__.py", line 51, in <module>
    raise ImportError("IfcOpenShell not built for '%s'" % python_distribution)
ImportError: IfcOpenShell not built for 'linux/64bit/python3.8'
```

It seems like there is still some issues with the ifcopenshell installation. Any thoughts?

**Comment by aothms** [comment]
> -- Looking for Open CASCADE library files in: /tmp/occ_dir/lib

Hm yeah likely the runtime linker cannot find the library files in that directory. I think (but I'm really not sure) you can move the occt libs you get from conda to some default paths like /usr/include and /usr/lib and have both cmake and the runtime linker find them automatically.

Since you're on ubuntu 20.04 you can also just install opencascade from apt https://packages.ubuntu.com/focal/libocct-data-exchange-dev Something like `libocct-data-exchange-dev libocct-foundation-dev libocct-modeling-algorithms-dev libocct-modeling-data-dev` should be sufficient I guess. The include dir will be picked up automatically https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/cmake/CMakeLists.txt#L234 but I think for the OCC_LIBRARY_DIR you have to specify `/usr/lib/x86_64-linux-gnu`

**Comment by Krande** [comment]
Okay, that did the trick!

I notice 1 test is failing (see [here](https://github.com/Krande/IfcOpenShell/runs/3333915545?check_suite_focus=true#step:7:8))

```
Traceback (most recent call last):
  File "src/ifcopenshell-python/ifcopenshell/test_ids.py", line 2, in <module>
    import ids
  File "/home/runner/work/IfcOpenShell/IfcOpenShell/src/ifcopenshell-python/ifcopenshell/ids.py", line 9, in <module>
    ids_schema = XMLSchema("http://standards.buildingsmart.org/IDS/ids.xsd")
  File "/usr/local/lib/python3.8/dist-packages/xmlschema/validators/schema.py", line 326, in __init__
    self.source = XMLResource(source, base_url, allow, defuse, timeout)
  File "/usr/local/lib/python3.8/dist-packages/xmlschema/resources.py", line 421, in __init__
    self.parse(source, lazy)
  File "/usr/local/lib/python3.8/dist-packages/xmlschema/resources.py", line 627, in parse
    with urlopen(url, timeout=self._timeout) as resource:
  File "/usr/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.8/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/usr/lib/python3.8/urllib/request.py", line 640, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.8/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found
Error: Process completed with exit code 1.
```

if I am not mistaken it is trying to open a http url (http://standards.buildingsmart.org/IDS/ids.xsd) that is no longer available?

**Comment by aothms** [comment]
Yes I talked to @atomczak about this and he will put that particular XSD def in the source repo so that the web request is no longer necessary?

So what are the next steps? It seems to me the current [CI action](https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/.github/workflows/ci.yml) doesn't really do CI. It's more CD. Should we rename it to something else and have yours in place of the current one triggered by both commit and pull request?

What are your thoughts?

**Comment by Krande** [comment]
Sure, I am fine with renaming the current ci.yml and create a new ci.yml based on what I have now. 

To start I can make a PR with just that.


Regarding the tests themselves maybe we should consider moving the existing tests into a pytest/unittests setup so that we can more easily add/organize tests into different categories and run the tests in parallel?

**Comment by aothms** [comment]
> Regarding the tests themselves maybe we should consider moving the existing tests into a pytest/unittests setup so that we can more easily add/organize tests into different categories and run the tests in parallel?

Yes definitely! We also need a lot more tests and not only python, but also C++. I prefer `unittests` as that's what @atomczak had used in his IDS tests, but I'm not up to date on the differences with pytest. If you want to advance on this that would be magnificent but I also think we need to as a community come together and see what we think of the current code quality as a whole and what needs to be done.

**Comment by Krande** [comment]
Unittests it is then:) 

I will create a PR of the ci.yml change now and after I can create another PR for a proposal of reorganizing the tests into a unittests setup.

**Comment by atomczak** [comment]
> if I am not mistaken it is trying to open a http url (http://standards.buildingsmart.org/IDS/ids.xsd) that is no longer available?

The failing test has been fixed by embedding the xsd file in ifcopenshell. It is in GSoC PR, not yet merged.

**Comment by Krande** [comment]
Okay, that sounds great @atomczak! 

I will hold off on creating a new PR (with a proposal for a unittest setup) after your PR (and my ci-PR) has been merged.

**Comment by Moult** [comment]
Is Travis still in use?
